### ISSUE TITLE
channelListController.synchronize multiple completion callbacks

### ISSUE THREAD

**Issue by AndrewSB** [GetStream/stream-chat-swift#2568]
## What did you do?

i'm trying to wrap `ChatChannelListController.synchronize` in a continuation, so i can make it async

```swift
        return await withCheckedContinuation { continuation in
            channelListController.synchronize { error in
                if let error = error {
                    CrashReporter.logNonFatal(error)
                } else {
                    self.channels = Array(channelListController.channels).filter { !$0.isGroupChat }
                }
                continuation.resume(returning: self.channels)
            }
        }
```

## What did you expect to happen?

i expected the callback to be called once, and only once, with either the channels or with an error 

## What happened instead?

when there's an error, the callback is called with an error first, then again after retry. this causes a swift runtime crash, 
`Fatal error: SWIFT TASK CONTINUATION MISUSE: fetchServerActiveChatChannels(currentUserId:) tried to resume its continuation more than once`

it looks like your `APIClient.operation` internally retries up to 3 times, calling completion upto 3 times. i'd expect not to receive an error in the completion of `synchronize(_ completion:)` if you're retrying.

thoughts?

**Comment by testableapple** [comment]
Hi @AndrewSB,
Thanks for reporting the issue. 

We are going to investigate this on our side and will get back to you as soon as possible.
In the meantime, you may try to erase the continuation once it is called (e.g.: https://forums.swift.org/t/how-to-prevent-swift-task-continuation-misuse/57581/2). Hope this will unblock you while we are investigating this.

Best,
Alexey

**Comment by testableapple** [comment]
@AndrewSB, could you share the SDK version, please? We fixed a couple of issues around this lately, so if you are not on the latest version, would be nice to try it out.

Best,
Alexey

**Comment by AndrewSB** [comment]
the bug seems really straightforward to me, your networking code calls a completion handler multiple times, one for each retry, if theres a network error.

if thats expected, it shouldn't be called a "completion" handler and satisfy the contract in some other way, i'd recommend just not calling the handler for any errors until the retries have been spent

**Comment by polqf** [comment]
Hi @AndrewSB, 
If you have a close look at the implementation, you would see how we are not calling the completion block when we are retrying.
Would be nice to understand the whole flow you execute for that to happen.

Is that consistently happening? Please do share the integration code, as I don't really know where this execution falls based on the snippet you shared some days ago ([this one](https://gist.github.com/AndrewSB/d1062b1912e6ea7d1db0147fd40a7dc7))

**Comment by AndrewSB** [comment]
just saw this again while in development

<img width="1656" alt="Screenshot 2023-05-08 at 11 41 23 AM" src="https://user-images.githubusercontent.com/3814772/236905679-313b0ed5-f57f-40cd-b690-e9444f0c68aa.png">

**Comment by AndrewSB** [comment]
on 4.31.0

**Comment by nuno-vieira** [comment]
More details from @AndrewSB here: https://github.com/GetStream/stream-chat-swift/issues/2627

**Comment by AndrewSB** [comment]
hey! this is our number one crash currently, it seems like if the request fails we try to `complete` with an error, then later if a retry succeeds we `complete` again with success

**Comment by polqf** [comment]
Hi @AndrewSB,

We just managed to reproduce it, and will be working on a fix. Sorry for the delay.

**Comment by AndrewSB** [comment]
appreciate it! happy you were able to find it

**Comment by nuno-vieira** [comment]
Hi @AndrewSB,

Just an update on this one. Pol is on Holiday, and since we are low in capacity right now, it might take a little bit more time to deliver this one. 

Hope you understand!

Best,
Nuno

**Comment by laevandus** [comment]
Hi @AndrewSB, 
I investigated this issue and tracked down the reason it was happening. It turned out to be the same problem what was fixed in [#2827](https://github.com/GetStream/stream-chat-swift/pull/2827). That fix was part of the 4.39.0 release. Please let us know if you can see this issue still or there have been other similar crashes when wrapping completion handler based code with async-await.

**Comment by AndrewSB** [comment]
thank you, i'll remove our workaround and let you know if there are still crashes
